### ISSUE TITLE
[HOLD for payment 2021-12-13] [HOLD for payment 2021-12-06] Delete comment - Deleted comment is still visible on LHN

### ISSUE THREAD

**Issue by isagoico** [Expensify/App#3293]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Expected Result:
LHN should show the message that was before the deleted message. 

## Actual Result:
LHN still shows the deleted comment. After refreshing the page, the email address of the sender is displayed. 

## Action Performed:
1. Log in to e.cash 
2. Navigate to a conversation
3. Send a message 
4. Delete the message
5. Check the LHN
6. Refresh the page

## Workaround:
No need, visual issue. 

## Platform:
Where is this issue occurring?

 Web ✔️ 
 iOS✔️
 Android ✔️ 
 Desktop App (Don't have a mac to check) 
 Mobile Web ✔️ 

**Version Number:** 1.0.59-0

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 

![image](https://user-images.githubusercontent.com/44479856/120511512-7af11f80-c398-11eb-9433-bc33d06700c7.png)

https://user-images.githubusercontent.com/44479856/120511373-5bf28d80-c398-11eb-8337-f96bbe508290.mp4

**Expensify/Expensify Issue URL:**

[view this job on upwork ](https://www.upwork.com/jobs/~01c6da4a31ffe3e961)

[View all open jobs on Upwork](https://www.upwork.com/ab/jobs/search/?q=Expensify%20React%20Native&sort=recency&user_location_match=2)

___ 
From @rafecolton https://expensify.slack.com/archives/C01GTK53T8Q/p1622591805431300

> Deleted comment shows as the recipient's email in the LHN. At first, the deleted comment still showed there, and then after a hard refresh, it just shows the email. Expected behavior is it would show the previous comment as though the deleted comment had never occurred.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @cead22 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jboniface (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @NikkiWines (`Exported`), see https://stackoverflow.com/c/expensify/questions/7972 for more details.

**Comment by dklymenk** [comment]
There are 2 issues here (last message shown is not updated after deleting it and after reload it displays sender email) that need to addressed separately, but I think a solution can be refactored into one new function partially.

## Addressing the refresh issue

It is caused by the fact that function `getSimplifiedReportObject`, that is responsible for converting API data into object that is stored locally, gets last message by simply getting the last object value. This is not correct anymore because apparently a delete comment action just changes it's text to an empty string and it is still served by BE as an actual message.

The reason it displays an email is as simple as it being a fallback in case message is empty.

So my proposal here is to update the way `lastReportAction` is found here:

https://github.com/Expensify/Expensify.cash/blob/8b979de2cf00deafbba85c5c139f6d78cd1fd3f2/src/libs/actions/Report.js#L148-L153

with something like this:
```js
    const reportActionKey = _.find(
        Object.keys(reportActionList).reverse(),
        key => reportActionList[key].message.html !== '',
    );
    const lastReportAction = reportActionList[reportActionKey];
```
This will make sure last message has text (is not deleted).

## Addressing the on delete issue

The root cause of this is the fact that data like `lastMessageText`, `lastMessageTimestamp` and `lastActorEmail` for a specific report are not updated on delete action. It is the way things work for `updateReportActionMessage`

https://github.com/Expensify/Expensify.cash/blob/8b979de2cf00deafbba85c5c139f6d78cd1fd3f2/src/libs/actions/Report.js#L488-L493

and `updateReportWithNewAction`

https://github.com/Expensify/Expensify.cash/blob/8b979de2cf00deafbba85c5c139f6d78cd1fd3f2/src/libs/actions/Report.js#L528-L534

but `deleteReportComment` function is missing that code, so I would need to implement something similar there:
```js
    ...
    
    Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`, reportActionsToMerge);

    // If this is the most recent message, update the lastMessage
    if (reportAction.sequenceNumber === reportMaxSequenceNumbers[reportID]) {
        #TODO figure out how to get reportActionList here
        const reportActionList = []; 
        const reportActionKey = _.find(
            Object.keys(reportActionList).reverse(),
            key => _.last(reportActionList[key].message).text !== '',
        );
        const lastReportAction = reportActionList[reportActionKey];
        const createTimestamp = lastReportAction ? lastReportAction.created : 0;
        const lastMessageTimestamp = moment.utc(createTimestamp).unix();
        const lastMessageText = lodashGet(lastReportAction, ['message', 'html'], '')
            .replace(/((<br[^>]*>)+)/gi, ' ')
            .replace(/(<([^>]+)>)/gi, '');
        const lastActorEmail = lodashGet(lastReportAction, 'accountEmail', '');
        Onyx.merge(`${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {
            lastMessageText,
            lastMessageTimestamp,
            lastActorEmail,
        });
    }
        
     ...
```

## Refactoring

I'm not exactly sure it would be worth it, but the code where I get last non-empty message text, timestamp and email can be refactored into a separate function `getLastReportActionDetails(reportActionList)` that would return an object with `lastMessageText`, `lastMessageTimestamp` and `lastActorEmail`. This function could be then used in both `getSimplifiedReportObject` and `deleteReportComment`.

## Summary

I am yet to figure out how to get `reportActionList` in `deleteReportComment` function and I would also need to test if using sequence numbers to determine last message is good enough now that deleted messages can still have them, but otherwise that's it.

P.S. What does LHN stand for? It is referenced in the issue, but I couldn't find a relevant meaning [online](https://acronyms.thefreedictionary.com/LHN). Judging by the context I would assume it's synonymous with Last Chat Message, but I better be sure. Thanks.

**Comment by jboniface** [comment]
@dklymenk LHN = left hand navigation

**Comment by NikkiWines** [comment]
@dklymenk couple of notes on your proposal: 

I think we could simplify your proposed `lastReportAction` logic by doing the following: 

```Javascript 
    const lastReportAction = !_.isEmpty(reportActionList)
        ? _.find([...reportActionList].reverse(), reportAction => reportAction.message.html !== '')
        : null;
``` 

It's also notable that IOU report actions don't have `message.html`, but your logic accounted for that already 👍 

Your logic for `deleteReportComment` looks good, though I agree getting the `reportActionList` is a major blocker for being able to implement what you've proposed. 

cc: @sketchydroide since you wrote a lot of the delete comment logic in Report.js

**Comment by dklymenk** [comment]
@NikkiWines 

I have given it another look and there should be no issues with doing the following inside `deleteReportComment` to get access to `reportActionList`.

```js
Onyx.connect({
    key: `${ONYXKEYS.COLLECTION.REPORT_ACTIONS}${reportID}`,
    callback: (reportActions) => {
        // TODO get last non-empty message and update the relevant data for report with reportID
    },
});
```

Additionally in my original proposal the code is presented in such a way that last message is updated regardless of the API response. So there is no case handling when this optimistic update needs to be reverted.

What I suggest to avoid this problem is not doing an optimistic update at all. To do this we run the "update last message" code only after we get 200 response from BE, since we are running async code anyway, it shouldn't a big deal. The code will be a lot cleaner and we don't have to think about potential race conditions.

If you insist on there being an optimistic update for last message, I can do that too. Since we already have last report action in that function, I can get the data from it and there will be no need for doing any Onyx.connect stuff the second time around. But I will need to make sure there is nothing funny happening like optimistic update occurring after the request response arrives.

**Comment by sketchydroide** [comment]
I think the approach @NikkiWines has is better, and I had thought of a similar solution.
as for the  `on delete` action, I wander if the regular `on edit` action suffers from similar problems, I mean when you edit a comment, the one in the LHN is not updated at the moment as well (at least I don't think it would be). 
Do you think this solution would work for both cases?

**Comment by dklymenk** [comment]
> I think the approach @NikkiWines has is better, and I had thought of a similar solution.

Yes, I agree the code for finding the last report action proposed by @NikkiWines is better.

> as for the on delete action, I wander if the regular on edit action suffers from similar problems, I mean when you edit a comment, the one in the LHN is not updated at the moment as well (at least I don't think it would be).
Do you think this solution would work for both cases?

My solution will only affect the `on delete` action, however, as I stated in my original proposal, `updateReportActionMessage` function is already updating the last message, so there is no such issue with `on edit`.

**Comment by sketchydroide** [comment]
sounds good then

**Comment by NikkiWines** [comment]
Cool, @jboniface can you hire @dklymenk for this job on Upwork please?

**Comment by dklymenk** [comment]
@NikkiWines What do you think about optimistic update? Should I bother with it or should I update the last message only after successful response from BE?

**Comment by NikkiWines** [comment]
@dklymenk we typically prefer an optimistic approach but it looks like `updateReportActionMessage` isn't being used optimistically at the moment. You can go with the simpler non-optimistic approach

**Comment by dklymenk** [comment]
@NikkiWines I have submitted a PR for this issue #3526 

There are a few issues I have run into on the way. They made me alter my original proposal in a few ways, so I thought I need to list them:
1. The callback for `Onyx.connect` gets called twice and sometimes even three times for me. I tried my best to figure out the cause, but after debugging `react-native-onyx` code I couldn't find the issue. Anyway, I used `_.once` to ensure the callback is called only once, and it seems to work perfectly.
2. I haven't stumbled upon any code worth refactoring into single function because `reportActionList` is a completely different data entity in context of `getSimplifiedReportObject` and `deleteReportComment` functions. In the first case it's an array of objects returned from API, and in the second one it's an object of objects from local storage. They are not 100% different, but IMO there are enough differences to avoid trying to write some generalized function that would extract `lastMessageDetails` from `reportActionList` in both cases.
3. The `if` statement from my original proposal, that was meant to ensure the code runs only if last message is deleted, works only for the first deleted message:
```
reportAction.sequenceNumber === reportMaxSequenceNumbers[reportID]
```
Updating `reportMaxSequenceNumbers` directly does not work from the context of that callback. If I was to do some `Onyx.merge` to update it, there would always be some weird issues when you delete 2 messages really fast and the `maxSequenceNumber` for the report wasn't updated fast enough, causing the second deletion to not pass the check. So, I decided to leave it as is. If you have any idea how this could be improved, I'd be happy to learn. Anyway, I don't believe this is a big deal.

Other than some minor additions for edge case handling (if you delete one and only message of the chat or next last message is an attachment), I think this is it.

Thanks.

**Comment by NikkiWines** [comment]
Hi, @dklymenk thanks for posting an update. 

1. I _think_ using `_.once` is fine here - though the multiple calls make me wonder if using `_.once` means you're not guaranteed to get the most up-to-date reportActionsList. 

2. I think it's fine without the refactoring 

3. I also don't think this is a huge deal - but I'll think on some potential alternatives

**Comment by dklymenk** [comment]
> I think using _.once is fine here - though the multiple calls make me wonder if using _.once means you're not guaranteed to get the most up-to-date reportActionsList.

In my testing that was never the case. I believe it's because every time we delete a comment a new instance of the onyx callback is called, so `_.once` function is called at least once on each comment deletion, so it does exactly what we want. 

Not sure if my explanation makes a lot of sense, but I think there is a way to refactor the code in such a way that would cause the target function to literally be executed only once. Let's say we declare the callback in another scope and then just pass the function returned from `_.once` as a callback to `Onyx.connect`. I believe that in that case the function would actually be called only once, preventing it from being called a second time at all.

**Comment by NikkiWines** [comment]
Hmm, I _think_ that's fine then... but just to be safe cc: @marcaaron just to be sure we're not going against anything by using `_.once` here.

**Comment by marcaaron** [comment]
Let's take it to `#expensify-open-source` ?

**Comment by dklymenk** [comment]
UPDATE: #3979 has been merged into main yesterday. It fixes this issue by showing 'comment deleted' message instead of previous one, if you delete the last comment. You are free to read discussion on #3743 to see how we got here, but basically showing next last message is either too inefficient or requires a lot of refactoring.

Thanks.

**Comment by NikkiWines** [comment]
That sounds like a good solution to this issue. Thanks for the update @dklymenk!

**Comment by dklymenk** [comment]
UPDATE: #3979 has been reverted by #4085 because it wasn't ready to be merged until #3737 is fixed. Otherwise it creates an issue like #4073 which is way more destructive for UX than #3737.

So I've created a new PR #4133 that is meant to be merged after #3737 is fixed. After that this issue can be considered closed.

Thanks.

**Comment by MelvinBot** [comment]
This issue has not been updated in over ***15*** days. @NikkiWines, @jboniface, @dklymenk eroding to ***Monthly*** issue.

P.S. Is everyone reading this sure this is really a near-term priority? Be brave: if you disagree, go ahead and close it out. If someone disagrees, they'll reopen it, and if they don't: one less thing to do!

**Comment by dklymenk** [comment]
UPDATE: A fix for #3737 is merged and deployed. So I was finally able to mark https://github.com/Expensify/App/pull/4133 as ready for review. :tada:

**Comment by kadiealexander** [comment]
Please refer to [this post](https://expensify.slack.com/archives/C01GTK53T8Q/p1633557259458600) for updated information on the `n6 hold`, we've raised the bonus to **$250** for all issues where a **PR is created** before the N6 hold is lifted.

**Comment by mallenexpensify** [comment]
`n6-hold` is lifted, label removed. @parasharrajat can you confirm we're still holding on https://github.com/Expensify/App/issues/5571?

**Comment by Beamanator** [comment]
Is this still being worked on? I'm still seeing this issue in dev & staging :'(

**Comment by dklymenk** [comment]
@Beamanator The PR has been merged and is deployed on staging.

I have just checked on staging and it looks good to me. Can you describe your issue in more detail?

**Comment by Beamanator** [comment]
Really? I tried this in dev & staging and got a similar message preview in the LHN. The only steps I took were:
1. Log in to New Expensify
2. Send message to account that exists but you've never messaged before
3. See the below

![Screen Shot 2021-11-23 at 5 04 21 PM](https://user-images.githubusercontent.com/3885503/143051079-60a681e4-eaca-48fc-9c56-be4614e48bf5.png)

**Comment by dklymenk** [comment]
Oh, I will look into it. You shouldn't see [Comment deleted] if that's the only message in the chat.

Previously it shown user's email, so I wonder if that's how it should be after my change or do we add some new placeholder.

@Beamanator  @NikkiWines 
Can I get a confirmation that you should indeed see user's email and not some other text when starting a new chat? I have missed this case in my proposals, testing and implementation.

**Comment by NikkiWines** [comment]
Yeah, I think if a user A deletes their first message to user B, we'd want to show the same information as we would if it was a new chat to User B and just show User B's email address.

**Comment by dklymenk** [comment]
Actually my question was about whether or not we should show email or something else when you start a new chat with a person, but your comment makes the scope even broader. I will consult the code and hopefully, come back with a solution that shows an email on both newly created chats and chats that had all their messages deleted.

**Comment by NikkiWines** [comment]
Sorry that my answer was unclear - I think we should show the email, not something else. But also feel free to dig into it further if you like!

**Comment by dklymenk** [comment]
Ok, so to fix the issue where it says "Comment deleted" when you create a new chat, I propose replacing
https://github.com/Expensify/App/blob/10d861328cbc2540be0ef20a9a904291a2a20e44/src/libs/actions/Report.js#L173-L178
with
```js
    let lastMessageText;
    if (report.reportActionCount > 0) {
        // We are removing any html tags from the message html since we cannot access the text version of any comments as
        // the report only has the raw reportActionList and not the processed version returned by Report_GetHistory
        // We convert the line-breaks in html to space ' ' before striping the tags
        lastMessageText = lastActionMessage
            .replace(/((<br[^>]*>)+)/gi, ' ')
            .replace(/(<([^>]+)>)/gi, '') || `[${translateLocal('common.deletedCommentMessage')}]`;
    } else {
        lastMessageText = null;
    }
```

This way if there are no report actions we default the LHN display to it's original behavior (showing email or phone).

Now when it comes to showing user's email if all the messages are deleted it gets more complicated. Since we don't have access to all reports in that function it would mean somehow fetching all the reports in an async manner, which as far as I am aware is something we are trying to avoid. This is especially the case for something that would run for each of your chats. I don't think this is feasible unless we get this info from BE, so I suggest just apply the fix above to remedy the obvious bug I have overseen in my original PR.

Thanks.

**Comment by NikkiWines** [comment]
That looks good to me @dklymenk

**Comment by dklymenk** [comment]
I have submitted the fix #6458.

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.16-10 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/6458
- https://github.com/Expensify/App/pull/6458


If no regressions arise, payment will be issued on 2021-12-06. :confetti_ball:

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.17-7 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/6458
- https://github.com/Expensify/App/pull/6458


If no regressions arise, payment will be issued on 2021-12-13. :confetti_ball:

**Comment by jboniface** [comment]
paid with bonus
